handler.go

```diff
@@ -1,13 +1,118 @@
 package metrics
 
 import (
+	"fmt"
 	"net/http"
 
 	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
 )
 
 // Handler returns the global http.Handler that provides the prometheus
-// metrics format on GET requests
+// metrics format on GET requests. This handler is no longer instrumented.
 func Handler() http.Handler {
-	return prometheus.Handler()
+	return promhttp.Handler()
+}
+
+// HttpHandlerOpts describes a set of configurable options of http metrics
+type HttpHandlerOpts struct {
+	Namespace          string
+	Subsystem          string
+	HandlerName        string
+	DurationBucket     []float64
+	RequestSizeBucket  []float64
+	ResponseSizeBucket []float64
+}
+
+var (
+	defaultDurationBucket     = []float64{0.05, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1., 2., 5., 10., 20., 30., 40., 50.}
+	defaultRequestSizeBucket  = prometheus.ExponentialBuckets(1024, 2, 22) //1K to 4G
+	defaultResponseSizeBucket = defaultRequestSizeBucket
+)
+
+func InstrumentHandler(handlerName string, handler http.Handler) http.HandlerFunc {
+	return InstrumentHandlerFunc(handlerName, handler.ServeHTTP)
+}
+
+func InstrumentHandlerFunc(handlerName string, handlerFunc func(http.ResponseWriter, *http.Request)) http.HandlerFunc {
+	return InstrumentHandlerFuncWithOpts(
+		HttpHandlerOpts{
+			Subsystem:   "http",
+			HandlerName: handlerName,
+		},
+		handlerFunc,
+	)
+}
+
+func InstrumentHandlerWithOpts(opts HttpHandlerOpts, handler http.Handler) http.HandlerFunc {
+	return InstrumentHandlerFuncWithOpts(opts, handler.ServeHTTP)
+}
+
+// InstrumentHandlerFuncWithOpts instruments http handler with in-flight gauge,
+// request count, handle duration, request size, and response size
+func InstrumentHandlerFuncWithOpts(opts HttpHandlerOpts, handlerFunc func(http.ResponseWriter, *http.Request)) http.HandlerFunc {
```

Comment:
This should take the handler type, not the function type. Also, it should return a handler.

Reply:
@stevvooe The deprecated functions uses http.HandlerFunc. I can added another wrapper that uses/returns a handler type. Do we need that ?

Reply:
I see. `http.HandlerFunc` is an implementation of `http.Handler`. I would generally only deal in `http.Handler` for exported types, but I can see this is a specific function. Doing it the other way around would be better, since it is odd to take a method value:

```
func InstrumentHandlerWithOpts(ops HTTPHandlerOpts, handler http.Handler) http.Handler {
// body of this function.
}

func InstrumentHandlerFuncWithOpts(opts HttpHandlerOpts, handlerFunc func(http.ResponseWriter, *http.Request)) http.Handler {
    return InstrumentHandlerWithOpts(opts, http.HandlerFunc(handlerFunc))
}
```